CUDADataFormats/Track/test/BuildFile.xml

```diff
@@ -0,0 +1,13 @@
+<use name="HeterogeneousCore/CUDAUtilities"/>
+
+<bin file="TrajectoryStateSOA_t.cpp" name="cpuTrajectoryStateSOA_t">
+  <use name="eigen"/>
+  <flags CXXFLAGS="-g -DGPU_DEBUG"/>
+</bin>
+
+<bin file="TrajectoryStateSOA_t.cu" name="gpuTrajectoryStateSOA_t">
```

Comment:
@fwyzard , can you please add cuda-gcc-support protection [a] for cuda released build products 

```
<iftool name="cuda-gcc-support">
....
....
</iftool>
```

Reply:
To be honest, I'm getting to the point where I will simply stop caring about the integration in CMSSW.

Reply:
I think the build system should figure out by itself.

Reply:
thats a good point @VinInn

Reply:
@smuzaffar , on architectures / compiler versions that do not support CUDA, like GCC 9, can we simply not build the CUDA externals, and automatically disable everything that depends on it ?

@makortel if we actually do this, would it be a reason to split the CUDA code into seprate libraries and plugins ?

Reply:
@fwyzard , we should keep on build cuda external otherwise how would we know if cuda support a GCC vet rsion or not :-)
But I agree that if cuda does not support our gcc version then we should disable the cuda based  build products in cmssw. This means we should have these products build in a way that they are optional . So cmssw should not fail if they do not exist.

Reply:
How all this will fit in "Heterogeneity"? In principle our ultimate goal is to have a single source.
The "cpu" version shall compile, link, be tested with any kind of compiler we wish to use.

Reply:
> we should keep on building cuda external otherwise how would we know if cuda support a GCC vet rsion or not :-)

Yes, well, what I meant is "if CUDA is not supported, building the external could fail".

However I realise that we do not really want to do that - for example, we can build llvm/clang with CUDA support as long as we have the libraries, even if GCC does not work with it.

So it's not 100% clear that `<use name="cuda"/>` should mean the CUDA libraries or the CUDA compiler.

Reply:
>  The "cpu" version shall compile, link, be tested with any kind of compiler we wish to use.

Yes, absolutely.

Reply:
> So it's not 100% clear that `<use name="cuda"/>` should mean the CUDA libraries or the CUDA compiler.

In principle the need of the CUDA compiler could be figured out if there are any `.cu` source files, right?

Reply:
Currently, yes.

If we use Alpaka or similar, we may end up with `.cc` files that need to be compiled by NVCC (or with `.cu` files that include `.cc` files).

Reply:
> @smuzaffar , on architectures / compiler versions that do not support CUDA, like GCC 9, can we simply not build the CUDA externals, and automatically disable everything that depends on it ?
> 
> @makortel if we actually do this, would it be a reason to split the CUDA code into seprate libraries and plugins ?

From the framework perspective not really (not sure if the split would make the life easier for the build system, ok libraries could make sense anyway). As outlined in https://github.com/cms-sw/cmssw/issues/28576#issuecomment-605681672, **if** we want a portable configuration, the framework needs to be able to construct all EDModules  and ESProducers for all the platforms the configuration supports. I.e. the CUDA EDProducers need to exist even when we can't build CUDA, implying that the pieces of code depending on CUDA needs to be protected with `#ifdef`s.

(not my favorite solution, but this is where we are today)

Reply:
> If we use Alpaka or similar, we may end up with `.cc` files that need to be compiled by NVCC (or with `.cu` files that include `.cc` files).

Right, but then the build system should be otherwise aware (either by explicit rules in the `BuildFile.xml`, or by built-in logic) whether to build for the CUDA backend or not.

Reply:
> the framework needs to be able to construct all EDModules and ESProducers for all the platforms the configuration supports. I.e. the CUDA EDProducers need to exist even when we can't build CUDA, implying that the pieces of code depending on CUDA needs to be protected with `#ifdef`s.

For what is worth, I really don't like this approach.

At this point, I think that the requirement to have unique hashes for the configuration is doing more harm to the process than (any ?) good.

Reply:
> > the framework needs to be able to construct all EDModules and ESProducers for all the platforms the configuration supports. I.e. the CUDA EDProducers need to exist even when we can't build CUDA, implying that the pieces of code depending on CUDA needs to be protected with `#ifdef`s.
> 
> For what is worth, I really don't like this approach.
> 
> At this point, I think that the requirement to have unique hashes for the configuration is doing more harm to the process than (any ?) good.

I replied in https://github.com/cms-sw/cmssw/issues/28576#issuecomment-610461957.

We can certainly discuss this subject in a Core Software meeting (today or later).